src/rustup-error/src/lib.rs

```diff
@@ -0,0 +1,494 @@
+//! To extend the error chain:
+//!
+//! ```
+//! use errors::ChainErr;
+//! try!(do_something().chain_err(|| "something went wrong"));
+//! ```
+//!
+//! `chain_err` can be called on any `Result` type where the contained
+//! error type implements `std::error::Error + Send + 'static`.  If
+//! the `Result` is an `Err` then `chain_err` evaluates the closure,
+//! which returns *some type that can be converted to `ErrorKind`*,
+//! boxes the original error to store as the cause, then returns a new
+//! error containing the original error.
+//!
+//! ## Foreign links
+//!
+//! Errors that do not conform to the same conventions as this library
+//! can still be included in the error chain. They are considered "foreign
+//! errors", and are declared using the `foreign_links` block of the
+//! `declare_errors!` macro. `Error`s are automatically created from
+//! foreign errors by the `try!` macro.
+//!
+//! Foreign links and regular links have one crucial difference:
+//! `From` conversions for regular links *do not introduce a new error
+//! into the error chain*, while conversions for foreign links *always
+//! introduce a new error into the error chain*. So for the example
+//! above all errors deriving from the `temp::Error` type will be
+//! presented to the user as a new `ErrorKind::Temp` variant, and the
+//! cause will be the original `temp::Error` error. In contrast, when
+//! `rustup_utils::Error` is converted to `Error` the two `ErrorKinds`
+//! are converted between each other to create a new `Error` but the
+//! old error is discarded; there is no "cause" created from the
+//! original error.
+//!
+//! ## Backtraces
+//!
+//! The earliest non-foreign error to be generated creates a single
+//! backtrace, which is passed through all `From` conversions and
+//! `chain_err` invocations of compatible types. To read the backtrace
+//! just call the `backtrace()` method.
+//!
+//! ## Iteration
+//!
+//! The `iter` method returns an iterator over the chain of error boxes.
+
+extern crate backtrace;
+
+use std::error::Error as StdError;
+use std::iter::Iterator;
+
+pub use backtrace::Backtrace;
+
+mod quick_error;
+
+#[macro_export]
+macro_rules! declare_errors {
+    (
+        types {
+            $error_name:ident, $error_kind_name:ident,
+            $chain_error_name:ident, $result_name:ident;
+        }
+
+        links {
+            $( $link_error_path:path, $link_kind_path:path, $link_variant:ident;  ) *
+        }
+
+        foreign_links {
+            $( $foreign_link_error_path:path, $foreign_link_variant:ident,
+               $foreign_link_desc:expr;  ) *
+        }
+
+        errors {
+            $( $error_chunks:tt ) *
+        }
+
+    ) => {
+
+
+        // The Error type
+        // --------------
+
+        #[derive(Debug)]
+        pub struct $error_name(pub $error_kind_name,
+                               pub Option<Box<::std::error::Error + Send>>,
+                               pub ::std::sync::Arc<$crate::Backtrace>);
```

Comment:
Ah, that is a good idea, though I might stuff them in a tuple to avoid having to name yet another type. I'd expect nobody to ever need to name that type, but instead to always match it with `_`.